### PR TITLE
persist the binary cache index in the misc cache and expose remote specs to `spack spec`

### PR DESCRIPTION
**Pull request by cosmicexplorer**
### Description
Closes #19085. Intends to solve the same problem.

### Changes
1. Add the test which uses `spack buildcache create` from #19085.

The PR has been split into several independent commits, with all of the work below in the final commit:

2. Make `spack buildcache list` use the directory `~/.spack/caches/indices` (using `config:misc_cache` as the root) to persist the `BinaryCacheIndex`.
2. Create `spec_index.py`:
    1. Move `InstallRecord` from `database.py`.
    1. Create `IndexQuery` class which performs the logic of `Database._query()`.
    1. Create a singleton `SpecIndex` to expose only local, only remote, or both local and remote specs for querying.
2. Make `spack spec`, `spack find`, and `spack install` concretize specs by referring to the appropriate `SpecIndex` instance.

### Result
The following works:
```bash
> spack buildcache list -l  # note hash for desired package is abcdefg 
> spack install /abcdefg    # if /abcdefg is from a remote buildcache
```

### CODE REVIEW

**Comment by cosmicexplorer** [comment]
@tgamblin This should be ready for review. The PR has been split into 4 independent commits, which makes it very easy for me to break out parts of it into pre-work PRs. I believe it should be clear upon inspection that the final commit is the only one that changes behavior, but it is very easy for me to split this up if needed.

Note that (as in the TODO section) that `spack buildcache list` and `spack install --cache-only` are not currently hooked up so that they work in the test. I am going to fix this before merging (there is a commented-out section of the new test case which uses them).

**Comment by cosmicexplorer** [comment]
Ok, the test should finally be passing now! 🎉

**Comment by cosmicexplorer** [comment]
Re @tgamblin @tldahlgren -- the workflow that this enables is also where it currently might cause a performance regression: in particular, when a hash prefix is used on the command line, spack will now unconditionally search all mirrors for that hash when parsing the spec. This is noted in two places:

1. https://github.com/spack/spack/blob/186d53b5ca878b855bcba2a605766fa950fd71ef/lib/spack/spack/spec.py#L4729-L4735
1. https://github.com/spack/spack/blob/186d53b5ca878b855bcba2a605766fa950fd71ef/lib/spack/spack/binary_distribution.py#L282-L292

Let me know if this is worth further investigation here. The perf regression would be in the case of `spack install asdf/xyz`, where `xyz` is the hash prefix. We *added* the ability for `xyz` to be in a remote db, but now we *unconditionally* search the remote db.

I think it would probably be good to ensure that `--no-cache` and `--cache-only` won't search both local and remote databases unconditionally. I'll try to implement that so users have a way to avoid taking the performance hit.

**Comment by cosmicexplorer** [comment]
Ok, I have just made changes to:
- `buildcache list`, `buildcache install` -- only look up from the remote spec index.
- `spec`, `solve` -- look up from the local *or* remote spec index.
- `install` -- look up from the local, remote, or merged local and remote spec index according to the values of `--use-cache`, `--no-cache`, and `--cache-only`.

File: lib/spack/llnl/util/lang.py [hunk 1]
```diff
@@ -193,15 +193,19 @@ def memoized(func):
     func.cache = {}
 
     @functools.wraps(func)
-    def _memoized_function(*args):
-        if not isinstance(args, Hashable):
+    def _memoized_function(*args, **kwargs):
+        assert isinstance(args, tuple), args
+        key = args + tuple(kwargs.items() if kwargs else [])
```
Comment (alalazo) [inline lib/spack/llnl/util/lang.py]: I think this generalization may introduce unexpected behavior. If I'm reading this correctly a function signature like:
```python
def foo(bar, baz):
    pass
```
would generate three different keys for these calls:
```python
foo(1, 2)  # key should be (1, 2)
foo(1, baz=2)  # key should be (1, ('baz, 2))
foo(bar=1, baz=2)  # key should be (('bar', 1), ('baz', 2))
```
which may be associated with three different values depending on the call. 

Also: we may extract this small change, if needed, into a small PR of its own.

**Review by alalazo — COMMENTED** [review]
It seems tests are hanging for some reason at:
```
lib/spack/spack/test/architecture.py::test_satisfy_strict_constraint_when_not_concrete[architecture_tuple1-constraint_tuple1]
```
where they wait to hit the wall-clock limit imposed by GA. I still need to check why that happens.

In the meanwhile I left a comment on the modification done to `memoized`.

File: lib/spack/llnl/util/lang.py (hunk 1 above)
Comment (cosmicexplorer) [reply lib/spack/llnl/util/lang.py]: Yes! Thank you. It has been reverted.

Comment (cosmicexplorer) [reply lib/spack/llnl/util/lang.py]: There is a battle-tested implementation that I wrote some parts of in `memo.py` in pants: https://github.com/pantsbuild/pants/blob/master/src/python/pants/util/memo.py I would borrow from that if we were to make another PR for this.

**Comment by cosmicexplorer** [comment]
> I still need to check why that happens.

I am pretty sure it was because of an unrelated change I made to `BoolValuedVariant` which I have since reverted. It was also causing `spack spec mpileaks%gcc^elfutils@01.70` to fail.

File: lib/spack/spack/cmd/__init__.py [hunk 2]
```diff
@@ -153,12 +153,13 @@ def parse_specs(args, **kwargs):
     concretize = kwargs.get('concretize', False)
     normalize = kwargs.get('normalize', False)
     tests = kwargs.get('tests', False)
+    index_location = kwargs.get('index_location', None)
```
Comment (tgamblin) [inline lib/spack/spack/cmd/__init__.py:156]: This shouldn't be exposed to `cmd/__init__.py` -- if the spec parser needs it, the spec parser should ask for it from the spec index.

File: lib/spack/spack/fetch_strategy.py [hunk 3]
```diff
@@ -509,6 +508,7 @@ def archive(self, destination):
         if not self.archive_file:
             raise NoArchiveFileError("Cannot call archive() before fetching.")
 
+        import spack.util.web as web_util
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py:511]: was this needed to avoid a circular import?

File: lib/spack/spack/variant.py [hunk 4]
```diff
@@ -122,6 +122,7 @@ def validate_or_raise(self, vspec, pkg=None):
         # Check and record the values that are not allowed
         not_allowed_values = [
             x for x in value
+            # TODO: avoid the multiple special-cases for '*' with a wrappper class.
```
Comment (tgamblin) [inline lib/spack/spack/variant.py:125]: this TODO doesn't have to do with this PR

File: lib/spack/spack/cmd/deprecate.py [hunk 5]
```diff
@@ -24,8 +24,8 @@
 import spack.cmd.common.arguments as arguments
 import spack.environment as ev
 
-from spack.error import SpackError
 from spack.database import InstallStatuses
+from spack.error import SpackError
```
Comment (tgamblin) [inline lib/spack/spack/cmd/deprecate.py:28]: no reason to do this in this PR. Remove.

File: lib/spack/spack/cmd/solve.py [hunk 6]
```diff
@@ -93,7 +94,8 @@ def solve(parser, args):
     if models < 0:
         tty.die("model count must be non-negative: %d")
 
-    specs = spack.cmd.parse_specs(args.specs)
+    merged = spack.spec_index.IndexLocation.LOCAL_AND_REMOTE()
+    specs = spack.cmd.parse_specs(args.specs, index_location=merged)
```
Comment (tgamblin) [inline lib/spack/spack/cmd/solve.py:98]: this PR should not affect every call to `parse_specs` - remove this and make the spec parser know about the index.  the index can learn about what potential places there are to get hashes by looking at spack's mirror configuration.

File: lib/spack/spack/cmd/spec.py [hunk 7]
```diff
@@ -88,16 +90,15 @@ def spec(parser, args):
...
+                kwargs['hashes'] = False  # Always False for input spec
+                print("Input spec")
+                print("--------------------------------")
+                print(spec.tree(**kwargs))
+
+                kwargs['hashes'] = args.long or args.very_long
+                print("Concretized")
+                print("--------------------------------")
+                spec.concretize()
+                print(spec.tree(**kwargs))
```
Comment (tgamblin) [inline lib/spack/spack/cmd/spec.py:104]: There is no reason to change the control flow here in this PR.  Revert this -- using `continue` was fine, right?

File: lib/spack/spack/cmd/ci.py [hunk 8]
```diff
@@ -66,15 +66,7 @@ def setup_parser(subparser):
...
-        default=False, help="""Spack always check specs against configured
-binary mirrors when generating the pipeline, regardless of whether or not
-DAG pruning is enabled.  This flag controls whether it might attempt to
-fetch remote spec.yaml files directly (ensuring no spec is rebuilt if it is
-present on the mirror), or whether it should reduce pipeline generation time
-by assuming all remote buildcache indices are up to date and only use those
-to determine whether a given spec is up to date on mirrors.  In the latter
-case, specs might be needlessly rebuilt if remote buildcache indices are out
-of date.""")
+        default=False, help="""Deprecated -- does nothing.""")
```
Comment (tgamblin) [inline lib/spack/spack/cmd/ci.py:69]: We should just remove the option if it's really no longer needed, but I think this is still needed.

I'm not 100% sure this should be deprecated.  The CI command shouldn't always check the index -- sometimes it loads the data for one specific (known) file in the mirror, by hash.  In our pipelines I think deprecating this will cause processes to fetch the whole index when they do not need to -- so we should talk about this one.

File: lib/spack/spack/cmd/uninstall.py [hunk 9]
```diff
@@ -65,6 +68,7 @@ def setup_parser(subparser):
 
 
 def find_matching_specs(env, specs, allow_multiple_matches=False, force=False):
+    # type: (ev.Environment, List[Spec], bool, bool) -> Iterable[Spec]
```
Comment (tgamblin) [inline lib/spack/spack/cmd/uninstall.py:71]: Let's just remove all the type information here.

File: lib/spack/spack/cmd/uninstall.py [hunk 10]
```diff
@@ -219,9 +227,14 @@ def do_uninstall(env, specs, force):
...
-    is_ready = lambda x: not spack.store.db.query_by_spec_hash(x)[1].ref_count
+    local_index = spack.spec_index.IndexLocation.LOCAL().spec_index_for()
+
+    def is_ready(x):
+        pfx = spack.spec_index.HashPrefix(x)
+        entry = local_index.lookup_ensuring_single_match(pfx)
+        return not entry.record.ref_count
     if force:
-        is_ready = lambda x: True
+        is_ready = lambda x: True                           # noqa
```
Comment (tgamblin) [inline lib/spack/spack/cmd/uninstall.py:237]: Did this really need noqa?

File: lib/spack/spack/cmd/uninstall.py [hunk 11]
```diff
@@ -82,30 +86,34 @@ def find_matching_specs(env, specs, allow_multiple_matches=False, force=False):
...
-        matching = spack.store.db.query_local(spec, hashes=hashes,
-                                              installed=install_query)
+    local_index = spack.spec_index.IndexLocation.LOCAL().spec_index_for()
+
+    install_query = [InstallStatuses.INSTALLED, InstallStatuses.DEPRECATED]
+    index_query = spack.spec_index.IndexQuery(
+        query_specs=specs, hashes=hashes, installed=install_query)
+
+    for spec, matching in local_index.query_collecting_result_map(index_query).items():
+        matching_specs = [cspec.spec for cspec in matching]
```
Comment (tgamblin) [inline lib/spack/spack/cmd/uninstall.py:96]: why change the name here?  It creates noise in the PR -- was there a naming conflict?

File: lib/spack/spack/installer.py [hunk 12]
```diff
@@ -573,6 +579,7 @@ def log(pkg):
 
 
 def package_id(pkg):
+    # type: (spack.package.PackageBase) -> Optional[str]
```
Comment (tgamblin) [inline lib/spack/spack/installer.py:582]: remove

File: lib/spack/spack/installer.py [hunk 13]
```diff
@@ -1431,7 +1437,9 @@ def install(self):
             # (i.e., if external or upstream) BUT flag it as installed since
             # some package likely depends on it.
             if not task.explicit:
-                if _handle_external_and_upstream(pkg, False):
+                if spec.concrete and not spec.external:
+                    pass
+                elif _handle_external_and_upstream(pkg, False):
```
Comment (tgamblin) [inline lib/spack/spack/installer.py:1442]: why is this change in this PR?

File: lib/spack/spack/installer.py [hunk 14]
```diff
@@ -1614,17 +1622,33 @@ def install(self):
                    request.install_args.get('install_package') and
                    request.pkg_id not in self.installed]
         if exists_errors or failed_explicits or missing:
+            all_error_messages = []
```
Comment (tgamblin) [inline lib/spack/spack/installer.py:1625]: does the change here have anything to do with this PR?

File: lib/spack/spack/concretize.py [hunk 15]
```diff
@@ -501,9 +502,11 @@ def _proper_compiler_style(cspec, aspec):
                 if _proper_compiler_style(c, spec.architecture)).copy()
         except StopIteration:
             # No compiler with a satisfactory spec has a suitable arch
+            assert spec.architecture is not None, spec
             _compiler_concretization_failure(
                 other_compiler, spec.architecture)
 
+        assert spec.compiler is not None, spec
```
Comment (tgamblin) [inline lib/spack/spack/concretize.py:509]: are the asserts here needed?  they seem unrelated to the PR.

File: lib/spack/spack/patch.py [hunk 16]
```diff
@@ -383,15 +383,16 @@ def update_package(self, pkg_fullname):
...
+        for sha256, package_to_patch in index.items():
             p2p = self.index.setdefault(sha256, {})
             p2p.update(package_to_patch)
 
     def update(self, other):
         """Update this cache with the contents of another."""
-        for sha256, package_to_patch in other.index.items():
-            p2p = self.index.setdefault(sha256, {})
-            p2p.update(package_to_patch)
+        self._update_index(other.index)
```
Comment (tgamblin) [inline lib/spack/spack/patch.py:395]: these changes seem very unrelated to the PR.

File: lib/spack/spack/spec.py [hunk 17]
```diff
@@ -1094,8 +1103,10 @@ def __init__(self, spec_like=None,
         # Build spec should be the actual build spec unless marked dirty.
         self._build_spec = None
 
+        if index_location is None:
+            index_location = spack.spec_index.IndexLocation.LOCAL()
         if isinstance(spec_like, six.string_types):
-            spec_list = SpecParser(self).parse(spec_like)
+            spec_list = SpecParser(self, index_location=index_location).parse(spec_like)
```
Comment (tgamblin) [inline lib/spack/spack/spec.py:1109]: I do not think that we should change the whole spec constructor just for this -- I think the index location should be hidden from users of `Spec`.

File: lib/spack/spack/database.py [hunk 18]
```diff
@@ -1461,25 +1476,30 @@ def _query(
...
                 results.append(rec.spec)
+                continue
 
-        return results
+            if isinstance(query_spec, (list, spack.spec.Spec)):
+                if isinstance(query_spec, spack.spec.Spec):
+                    query_spec = [query_spec]
+                recorded_spec = rec.spec
+                if for_all_architectures:
+                    recorded_spec = recorded_spec.without_architecture_or_compiler()
```
Comment (tgamblin) [inline lib/spack/spack/database.py:1488]: `--allarch` is not a concern that we should be letting into `database.py`.  It's really a UI concern, not a core database concern.  If the user wants to see only some architectures, they should filter results outside the DB or we should make a better query object.  We shouldn't have a specific for_all_architectures argument here.

File: lib/spack/spack/database.py [hunk 19]
```diff
@@ -288,7 +289,14 @@ def __getattribute__(self, name):
         """
 
 
-class Database(object):
+def _query_docvar(f):
+    if f.__doc__ is None:
+        f.__doc__ = ""
+    f.__doc__ += _query_docstring
+    return f
```
Comment (tgamblin) [inline lib/spack/spack/database.py:296]: this is cool but do we need it in this PR?

File: lib/spack/spack/database.py [hunk 20]
```diff
@@ -701,18 +709,22 @@ def _assign_dependencies(self, hash_key, installs, data):
 
                 spec._add_dependency(child, dtypes)
 
-    def _read_from_file(self, filename):
+    def fill_from_file(self, filename):
```
Comment (tgamblin) [inline lib/spack/spack/database.py:712]: Why is the name changed here?  I guess you needed to makem the json one public, and it makes sense to make this one public too.  That's fine but I'd call them `read_from_file` and `read_from_json`.

**Review by tgamblin — CHANGES_REQUESTED** [review]
This is the first half of the review -- mostly about what probably shouldn't be in this PR... moving on to the parts that should :)

File: lib/spack/spack/spec_index.py [hunk 21]
```diff
@@ -0,0 +1,323 @@
...
+
+from six import add_metaclass
+
+from llnl.util.lang import memoized
+
+import spack.database
+import spack.spec
+
+
+class ConcretizedSpec(object):
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:18]: Not clear to me why we need this class.  Is the existing Spec API bad for this?  I don't see anything on here that we can't do with `Spec`.

File: lib/spack/spack/spec_index.py [hunk 22]
```diff
@@ -0,0 +1,323 @@
...
+    def spec_string_name_hash_only(self):
+        # type: () -> str
+        return '{0}/{1}'.format(self.spec.name, self.into_hash().complete_hash)
+
+    def __repr__(self):
+        return '{0}(spec={1!r})'.format(type(self).__name__, self.spec)
+
+
+class IndexEntry(object):
+    concretized_spec = None  # type: ConcretizedSpec
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:53]: I think this should just be `Spec`

File: lib/spack/spack/spec_index.py [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+    record = None            # type: spack.database.InstallRecord
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:54]: This is odd to me b/c the record should already have a spec.

File: lib/spack/spack/spec_index.py [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+        self.record = record
+
+    def __repr__(self):
+        return '{0}(concretized_spec={1!r}, record={2!r})'.format(
+            type(self).__name__, self.concretized_spec, self.record)
+
+
+@add_metaclass(ABCMeta)
+class PartialHash(object):
+    FULL_HASH_STRING_LENGTH = 32  # type: ClassVar[int]
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:68]: would rather not have this defined in multiple places.

File: lib/spack/spack/spec_index.py [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+
+
+class CompleteHash(PartialHash):
+    @abstractproperty
+    def complete_hash(self):
+        # type: () -> str
+        pass
+
+
+class ConcreteHash(CompleteHash):
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:113]: All the hash types here are losing me -- do we need them?  I think hashes can be much simpler than this...

File: lib/spack/spack/spec_index.py [hunk 26]
```diff
@@ -0,0 +1,323 @@
...
+            end_date=self.end_date,
+            hashes=self.hashes,
+            for_all_architectures=(False if self.for_all_architectures is None
+                                   else self.for_all_architectures),
+        )
+
+
+@add_metaclass(ABCMeta)
+class SpecIndexable(object):
+    """Interface to look up and traverse known specs from multiple sources."""
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:180]: we don't need to make interface types in Spack -- we tend to stay away from them.

File: lib/spack/spack/spec_index.py [hunk 27]
```diff
@@ -0,0 +1,323 @@
...
+        # type: (PartialHash) -> Iterator[IndexEntry]
+        pass
+
+    @abstractmethod
+    def spec_index_query(self, query):
+        # type: (IndexQuery) -> Iterator[ConcretizedSpec]
+        pass
+
+
+class SpecIndex(object):
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:193]: right now this is IMO overly complex -- it needs to be refactored to just take a list of DB's, and to look up hashes in those.  Where the hashes come from should be someone else's concern.

the local DB is a DB, the remote DB is a DB, etc. -- they're all the same.  I don't think we need separate methods or anything for each kind of DB.  I was expecting that this would just look up *hashes* and map them to specs, from several DBs.

File: lib/spack/spack/spec_index.py [hunk 28]
```diff
@@ -0,0 +1,323 @@
...
+                if concretized_spec.spec.satisfies(spec, strict=True):
+                    spec_to_results[spec].append(concretized_spec)
+
+        return spec_to_results
+
+    def __repr__(self):
+        return '{0}(indices={1!r})'.format(type(self).__name__, self.indices)
+
+
+class IndexLocation(object):
```
Comment (tgamblin) [inline lib/spack/spack/spec_index.py:279]: this can get removed -- we shouldn't need to care about the index location.

File: lib/spack/spack/binary_distribution.py [hunk 29]
```diff
@@ -48,409 +52,229 @@
...
-    up, but cache misses shouldn't break any spack functionality.
 
-    At the moment, everything in this class is initialized as lazily as
-    possible, so that it avoids slowing anything in spack down until
-    absolutely necessary.
+def _binary_index_location():
+    """Share a merged BinaryCacheIndex for remote buildcaches, in the user's homedir."""
+    cache_root = os.path.join(misc_cache_location(), 'indices')
+    cache_root = spack.util.path.canonicalize_path(cache_root)
+    return cache_root
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py:64]: this is good.

File: lib/spack/spack/binary_distribution.py [hunk 30]
```diff
@@ -48,409 +52,229 @@
...
+    except (URLError, web_util.SpackWebError) as e:
+        raise WebFailure('Unable to read index {0}'.format(index_fetch_url), e)
 
-        This method does not trigger reading anything from remote mirrors, but
-        rather just checks if the concrete spec is found within the cache.
+    locally_computed_hash = compute_hash(index_object_str)
 
-        The cache can be updated by calling ``update()`` on the cache.
+    if locally_computed_hash != expect_hash:
+        # Don't forget this could be sabotage!
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py:116]: Not getting why this could be sabotage

File: lib/spack/spack/binary_distribution.py [hunk 31]
```diff
@@ -48,409 +52,229 @@
...
+        raise InvalidIndexHash(expect_hash, locally_computed_hash,
+                               'remote index at {0}'.format(mirror_url),
+                               'this indicates an error in index tranmission')
+        # tty.error(msg_tmpl.format(locally_computed_hash, expect_hash))
```
Comment (tgamblin) [inline lib/spack/spack/binary_distribution.py:120]: this is commented out

**Review by tgamblin — CHANGES_REQUESTED** [review]
ok here are some more comments.

**Comment by cosmicexplorer** [comment]
Going to break this out into multiple other PRs despite as well as because of the immense effort required to get it passing CI.

**Comment by cosmicexplorer** [comment]
Reopening this so I don't forget about it.

**Comment by alalazo** [comment]
Closing, since there has been no activity in a while. Feel free to reopen if you want to continue work on this.
